hordelib/nodes/comfyui_layerdiffuse/layered_diffusion.py

```diff
@@ -0,0 +1,659 @@
+                        : weight.shape[1],
+                        : weight.shape[2],
+                        : weight.shape[3],
+                    ] = weight
+                    new_weight[
+                        : new_diff.shape[0],
+                        : new_diff.shape[1],
+                        : new_diff.shape[2],
+                        : new_diff.shape[3],
+                    ] += new_diff
+                    new_weight = new_weight.contiguous().clone()
+                    weight = new_weight
+        return weight
+
+    return calculate_weight
+
+
+ModelPatcher.calculate_weight = calculate_weight_adjust_channel(ModelPatcher.calculate_weight)
+
+# ------------ End patching ComfyUI ------------
+
+
+class LayeredDiffusionDecode:
+    """
+    Decode alpha channel value from pixel value.
+    [B, C=3, H, W] => [B, C=4, H, W]
+    Outputs RGB image + Alpha mask.
+    """
+
+    @classmethod
+    def INPUT_TYPES(s):
+        return {
+            "required": {
+                "samples": ("LATENT",),
+                "images": ("IMAGE",),
+                "sd_version": (
+                    [
+                        StableDiffusionVersion.SD1x.value,
+                        StableDiffusionVersion.SDXL.value,
+                    ],
+                    {
+                        "default": StableDiffusionVersion.SDXL.value,
+                    },
+                ),
+                "sub_batch_size": (
+                    "INT",
+                    {"default": 16, "min": 1, "max": 4096, "step": 1},
+                ),
+            },
+        }
+
+    RETURN_TYPES = ("IMAGE", "MASK")
+    FUNCTION = "decode"
+    CATEGORY = "layer_diffuse"
+
+    def __init__(self) -> None:
+        self.vae_transparent_decoder = {}
+
+    def decode(self, samples, images, sd_version: str, sub_batch_size: int):
+        """
+        sub_batch_size: How many images to decode in a single pass.
+        See https://github.com/huchenlei/ComfyUI-layerdiffuse/pull/4 for more
+        context.
+        """
+        sd_version = StableDiffusionVersion(sd_version)
+        if sd_version == StableDiffusionVersion.SD1x:
+            url = "https://huggingface.co/LayerDiffusion/layerdiffusion-v1/resolve/main/layer_sd15_vae_transparent_decoder.safetensors"
+            file_name = "layer_sd15_vae_transparent_decoder.safetensors"
+        elif sd_version == StableDiffusionVersion.SDXL:
+            url = "https://huggingface.co/LayerDiffusion/layerdiffusion-v1/resolve/main/vae_transparent_decoder.safetensors"
+            file_name = "vae_transparent_decoder.safetensors"
+
+        if not self.vae_transparent_decoder.get(sd_version):
+            model_path = load_file_from_url(url=url, model_dir=layer_model_root, file_name=file_name)
+            self.vae_transparent_decoder[sd_version] = TransparentVAEDecoder(
+                load_torch_file(model_path),
+                device=comfy.model_management.get_torch_device(),
```

Comment:
This node loading 3gb(!) worth of model outside of the usual comfy memory management scheme presents a challenge for ensuring it is cleaned up or otherwise handled in a <=12gb VRAM use case. I haven't rigorously explored what the implications are but If forsee this being a tripping point.

Reply:
Stress tested the model loading on my worker and I don't see any memory leaks with 4 simultaneous requests coming for transparent gens